Yeah, it could kill more then it should:

```
Killing process group 1
Processes in process group 1:
1 /bin/bash /run.sh
276 ./minio server --address :11111 ./minio_data
4391 gdb -batch -command script.gdb -p 3556
4392 /usr/bin/perl /usr/bin/ts %Y-%m-%d %H:%M:%S
4443 python3 /usr/bin/stress --hung-check --drop-databases --output-folder test_output --skip-func-tests  --global-time-limit 1200
4460 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_1  --client-option join_use_nulls=1 join_algorithm='parallel_hash' memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
4462 python3 /usr/bin/clickhouse-test --order=random --database=test_1 --client-option join_use_nulls=1 join_algorithm=parallel_hash memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
4495 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option join_use_nulls=1 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
4500 python3 /usr/bin/clickhouse-test --order=random --client-option join_use_nulls=1 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200
4530 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
4534 python3 /usr/bin/clickhouse-test --order=random --client-option memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
4550 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_7  --client-option join_use_nulls=1 join_algorithm='grace_hash' group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
4552 python3 /usr/bin/clickhouse-test --order=random --database=test_7 --client-option join_use_nulls=1 join_algorithm=grace_hash group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200
4643 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option implicit_transaction=1 throw_on_unsupported_query_inside_transaction=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
4648 python3 /usr/bin/clickhouse-test --order=random --client-option implicit_transaction=1 throw_on_unsupported_query_inside_transaction=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200
4772 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option join_use_nulls=1 memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
4782 python3 /usr/bin/clickhouse-test --order=random --client-option join_use_nulls=1 memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200
4812 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option optimize_trivial_approximate_count_query=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
4825 python3 /usr/bin/clickhouse-test --order=random --client-option optimize_trivial_approximate_count_query=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
4924 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_21  --client-option join_algorithm='parallel_hash' memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
4947 python3 /usr/bin/clickhouse-test --order=random --database=test_21 --client-option join_algorithm=parallel_hash memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
5033 /bin/sh -c /usr/bin/clickhouse-test --order=random  --client-option http_make_head_request=0 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
5036 python3 /usr/bin/clickhouse-test --order=random --client-option http_make_head_request=0 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
5054 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_25  --client-option join_use_nulls=1 join_algorithm='full_sorting_merge' group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
5060 python3 /usr/bin/clickhouse-test --order=random --database=test_25 --client-option join_use_nulls=1 join_algorithm=full_sorting_merge group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200
5177 /bin/sh -c /usr/bin/clickhouse-test --order=random --db-engine="Replicated('/test/db/test_29', 's1', 'r1')"  --database=test_29  --client-option enable_deflate_qpl_codec=1 enable_zstd_qat_codec=1 join_algorithm='auto' max_rows_in_join=1000 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
5180 python3 /usr/bin/clickhouse-test --order=random --db-engine=Replicated('/test/db/test_29', 's1', 'r1') --database=test_29 --client-option enable_deflate_qpl_codec=1 enable_zstd_qat_codec=1 join_algorithm=auto max_rows_in_join=1000 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200
5261 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_31  --client-option join_use_nulls=1 join_algorithm='parallel_hash' memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
5268 python3 /usr/bin/clickhouse-test --order=random --database=test_31 --client-option join_use_nulls=1 join_algorithm=parallel_hash memory_tracker_fault_probability=0.001 merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0.05 group_by_use_nulls=1 ignore_drop_queries_probability=0.5 --global_time_limit=1200
5289 /bin/sh -c /usr/bin/clickhouse-test --order=random --db-engine="Replicated('/test/db/test_32', 's1', 'r1')"  --client-option enable_deflate_qpl_codec=1 enable_zstd_qat_codec=1 optimize_trivial_approximate_count_query=1 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200  
5305 python3 /usr/bin/clickhouse-test --order=random --db-engine=Replicated('/test/db/test_32', 's1', 'r1') --client-option enable_deflate_qpl_codec=1 enable_zstd_qat_codec=1 optimize_trivial_approximate_count_query=1 http_make_head_request=0 ignore_drop_queries_probability=0.5 --global_time_limit=1200
5342 /bin/sh -c /usr/bin/clickhouse-test --order=random  --database=test_33  --client-option join_algorithm='partial_merge' group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas='parallel_replicas' parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200  
5349 python3 /usr/bin/clickhouse-test --order=random --database=test_33 --client-option join_algorithm=partial_merge group_by_use_nulls=1 ignore_drop_queries_probability=0.5 allow_experimental_parallel_reading_from_replicas=1 max_parallel_replicas=3 cluster_for_parallel_replicas=parallel_replicas parallel_replicas_for_non_replicated_merge_tree=1 --global_time_limit=1200
Process group 1 should be killed
Traceback (most recent call last):
  File "/usr/bin/clickhouse-test", line 3599, in <module>
    main(args)
  File "/usr/bin/clickhouse-test", line 2957, in main
    total_tests_run += do_run_tests(args.jobs, test_suite)
  File "/usr/bin/clickhouse-test", line 2592, in do_run_tests
    run_tests_array(
  File "/usr/bin/clickhouse-test", line 2302, in run_tests_array
    stop_tests()
  File "/usr/bin/clickhouse-test", line 429, in stop_tests
    cleanup_child_processes(os.getpid())
  File "/usr/bin/clickhouse-test", line 417, in cleanup_child_processes
    child_pgid = os.getpgid(child)
ProcessLookupError: [Errno 3] No such process
```

Hence CI fails

--- a/docker/test/stateful/run.sh
+++ b/docker/test/stateful/run.sh
@@ -232,15 +232,25 @@ function run_tests()
 
     set +e
 
+    TEST_ARGS=(
+        -j 2
+        --testname
+        --shard
+        --zookeeper
+        --check-zookeeper-session
+        --no-stateless
+        --hung-check
+        --print-time
+        "${ADDITIONAL_OPTIONS[@]}"
+        "$SKIP_TESTS_OPTION"
+    )
     if [[ -n "$USE_PARALLEL_REPLICAS" ]] && [[ "$USE_PARALLEL_REPLICAS" -eq 1 ]]; then
-        clickhouse-test --client="clickhouse-client --allow_experimental_parallel_reading_from_replicas=1 --parallel_replicas_for_non_replicated_merge_tree=1 \
-            --max_parallel_replicas=100 --cluster_for_parallel_replicas='parallel_replicas'" \
-            -j 2 --testname --shard --zookeeper --check-zookeeper-session --no-stateless --no-parallel-replicas --hung-check --print-time "${ADDITIONAL_OPTIONS[@]}" \
-        "$SKIP_TESTS_OPTION" 2>&1 | ts '%Y-%m-%d %H:%M:%S' | tee test_output/test_result.txt
-    else
-        clickhouse-test -j 2 --testname --shard --zookeeper --check-zookeeper-session --no-stateless --hung-check --print-time "${ADDITIONAL_OPTIONS[@]}" \
-        "$SKIP_TESTS_OPTION" 2>&1 | ts '%Y-%m-%d %H:%M:%S' | tee test_output/test_result.txt
+        TEST_ARGS+=(
+            --client="clickhouse-client --allow_experimental_parallel_reading_from_replicas=1 --parallel_replicas_for_non_replicated_merge_tree=1 --max_parallel_replicas=100 --cluster_for_parallel_replicas='parallel_replicas'"
+            --no-parallel-replicas
+        )
     fi
+    clickhouse-test "${TEST_ARGS[@]}" 2>&1 | ts '%Y-%m-%d %H:%M:%S' | tee test_output/test_result.txt
     set -e
 }
 

--- a/docker/test/stateless/run.sh
+++ b/docker/test/stateless/run.sh
@@ -264,11 +264,21 @@ function run_tests()
     TIMEOUT=$((MAX_RUN_TIME - 800 > 8400 ? 8400 : MAX_RUN_TIME - 800))
     START_TIME=${SECONDS}
     set +e
-    timeout --preserve-status --signal TERM --kill-after 60m ${TIMEOUT}s \
-        clickhouse-test --testname --shard --zookeeper --check-zookeeper-session --hung-check --print-time \
-            --no-drop-if-fail --test-runs "$NUM_TRIES" "${ADDITIONAL_OPTIONS[@]}" 2>&1 \
-    | ts '%Y-%m-%d %H:%M:%S' \
-    | tee -a test_output/test_result.txt
+
+    TEST_ARGS=(
+        --testname
+        --shard
+        --zookeeper
+        --check-zookeeper-session
+        --hung-check
+        --print-time
+        --no-drop-if-fail
+        --test-runs "$NUM_TRIES"
+        "${ADDITIONAL_OPTIONS[@]}"
+    )
+    timeout --preserve-status --signal TERM --kill-after 60m ${TIMEOUT}s clickhouse-test "${TEST_ARGS[@]}" 2>&1 \
+        | ts '%Y-%m-%d %H:%M:%S' \
+        | tee -a test_output/test_result.txt
     set -e
     DURATION=$((SECONDS - START_TIME))
 

--- a/src/Common/SignalHandlers.cpp
+++ b/src/Common/SignalHandlers.cpp
@@ -629,6 +629,7 @@ void HandledSignals::setupTerminateHandler()
 void HandledSignals::setupCommonDeadlySignalHandlers()
 {
     /// SIGTSTP is added for debugging purposes. To output a stack trace of any running thread at anytime.
+    /// NOTE: that it is also used by clickhouse-test wrapper
     addSignalHandler({SIGABRT, SIGSEGV, SIGILL, SIGBUS, SIGSYS, SIGFPE, SIGPIPE, SIGTSTP, SIGTRAP}, signalHandler, true);
 
 #if defined(SANITIZER)

--- a/tests/clickhouse-test
+++ b/tests/clickhouse-test
@@ -368,6 +368,13 @@ def kill_process_group(pgid):
         end="",
     )
     try:
+        # Let's try to dump stacktrace in client (useful to catch issues there)
+        os.killpg(pgid, signal.SIGTSTP)
+        # Wait some time for clickhouse utilities to gather stacktrace
+        if RELEASE_BUILD:
+            sleep(0.5)
+        else:
+            sleep(5)
         # NOTE: this still may leave some processes, that had been
         # created by timeout(1), since it also creates new process
         # group. But this should not be a problem with default
@@ -380,9 +387,8 @@ def kill_process_group(pgid):
         # ability to terminate childrens (though not always since
         # signals are asynchronous).
         os.killpg(pgid, signal.SIGTERM)
-        # This may not be enough, but this is at least something
-        # (and anyway it is OK to spend 0.1 second more in case of
-        # test timeout).
+        # We need minimal delay to let processes handle SIGTERM - 0.1 (this may
+        # not be enough, but at least something)
         sleep(0.1)
         os.killpg(pgid, signal.SIGKILL)
     except OSError as e:
@@ -2396,7 +2402,13 @@ class BuildFlags:
     POLYMORPHIC_PARTS = "polymorphic-parts"
 
 
+# Release and non-sanitizer build
+RELEASE_BUILD = False
+
+
 def collect_build_flags(args):
+    global RELEASE_BUILD
+
     result = []
 
     value = clickhouse_execute(
@@ -2420,6 +2432,8 @@ def collect_build_flags(args):
         result.append(BuildFlags.DEBUG)
     elif b"RelWithDebInfo" in value or b"Release" in value:
         result.append(BuildFlags.RELEASE)
+
+    RELEASE_BUILD = result == [BuildFlags.RELEASE]
 
     value = clickhouse_execute(
         args,

--- a/tests/clickhouse-test
+++ b/tests/clickhouse-test
@@ -374,7 +374,7 @@ def kill_process_group(pgid):
         if RELEASE_BUILD:
             sleep(0.5)
         else:
-            sleep(5)
+            sleep(10)
         # NOTE: this still may leave some processes, that had been
         # created by timeout(1), since it also creates new process
         # group. But this should not be a problem with default

